### DISCUSSION TITLE
What should be the standard for NETRC file?

### DISCUSSION THREAD

**Discussion by remidebette** [curl/curl#15713] [General]
Hi,

There has been a discussion about `.netrc` in the `uv` package which is gaining traction in the python ecosystem.
netrc can be used as a way to declare tokens for private packages.

The link to the discussion is here: https://github.com/astral-sh/uv/issues/9576

In a nutshell, `uv` allows for an environment variable named `NETRC` to define a path to an arbitrary `.netrc` file, overriding the home folder.

This environment variable comes from the netrc rust crate: https://github.com/gribouille/netrc/blob/f8b614440a8d45cae8a5bd90a43266d8f487b93b/src/lib.rs#L85-L105

Now, `uv` ensures `NETRC` is applied for **pypi style** packages urls, but one can also use **git style** urls. 
(thank you github for not supporting python private packages: https://github.com/github/roadmap/issues/94#issuecomment-1581086839)

As it so happens, such a `NETRC` env var does not work in the case of git urls, and according to this [discussion](https://lore.kernel.org/git/20241202211156.GG776185@coredump.intra.peff.net/) on the git mailing list, git uses libcurl with netrc enabled but without more configuration, so it does not use CURLOPT_NETRC_FILE as described here https://curl.se/libcurl/c/CURLOPT_NETRC_FILE.html

Now the question is how could this be harmonized? 
Could libcurl accept the env var?

**Comment by bagder** [comment]
Let me first remind all readers that .netrc is also not working the same everywhere, so there might be dragons when trying to use the same .netrc with multiple clients. I have blogged about that in the past: https://daniel.haxx.se/blog/2022/05/31/netrc-pains/

> Could libcurl accept the env var?

I vote no. A) This is not a "standard" env variable and B) env variables make for lousy APIs.

To me, its sounds like you primarily want git to work with the env var. It could fetch it and set the necessary libcurl option.

**Reply by zanieb** [reply]
For my own future knowledge, what makes an environment variable "standard"?

(moved this comment after realizing we're in GitHub "Discussions")

**Reply by bagder** [reply]
I don't mean to imply that this is an established term, but I would say that it is an environment variable that has been used for a specific purpose for a while, is widely understood and is supported by multiple independent software components (in an at least mostly coherent and agreed to way).
